src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -241,35 +241,51 @@ await foreach (ConfigurationSetting setting in _client.GetConfigurationSettingsA
 
                 // Block current thread for the initial load of key-values registered for refresh that are not already loaded
                 await Task.Run(() => LoadKeyValuesRegisteredForRefresh(data).ConfigureAwait(false).GetAwaiter().GetResult()).ConfigureAwait(false);
+
+                if (_options.OfflineCache != null)
+                {
+                    _options.OfflineCache.Export(_options, JsonSerializer.Serialize(data));
+                }
+
+                isLoadAllComplete = true;
             }
             catch (Exception exception) when (exception is RequestFailedException ||
                                               ((exception as AggregateException)?.InnerExceptions?.All(e => e is RequestFailedException) ?? false) ||
                                               exception is OperationCanceledException)
             {
                 if (_options.OfflineCache != null)
                 {
-                    data = JsonSerializer.Deserialize<IDictionary<string, ConfigurationSetting>>(_options.OfflineCache.Import(_options));
-
-                    if (data != null)
+                    // During startup or refreshAll scenario, we'll try to populate config from offline cache, if available
+                    string cachedData = _options.OfflineCache.Import(_options);
+                    if (cachedData != null)
                     {
-                        await SetData(data, ignoreFailures).ConfigureAwait(false);
-                        return;
+                        data = JsonSerializer.Deserialize<IDictionary<string, ConfigurationSetting>>(cachedData);
+                        isLoadAllComplete = true;
                     }
                 }
 
-                if (!ignoreFailures)
+                // If we're unable to load data from offline cache, check if we need to ignore or rethrow the exception 
+                if (!isLoadAllComplete && !ignoreFailures)
                 {
                     throw;
                 }
-
-                return;
             }
 
-            await SetData(data, ignoreFailures).ConfigureAwait(false);
-
-            if (_options.OfflineCache != null)
+            if (isLoadAllComplete)

```

Comment:
After thinking more about the refactoring, I think it makes sense to use a boolean flag to track if we were able to load the settings (from AppConfig or OfflineCache) in order to remove the return statements from the catch block. In my opinion, not using the flag would complicate this part of code and make it harder to understand next time.

Reply:
Did you consider initializing data to null and only setting it to non null in the places you are setting isLoadAllComplete to true? Then that would take the place of the bool.

Reply:
Yes, but Abhilash pointed out (and I agree) that we would essentially be using `data` to do the work of a bool flag, and change the intended responsibility of `data`, which would make the code less readable. Having an explicit bool flag makes the functionality more obvious.

Reply:
I won't block on it, but I don't agree in the necessity of introducing a new variable to track the state of a method. Always that will make me think something is wrong with the design. I would say that the pattern below would only be more complicated with the introduction of a new variable to track. 

```

object some = null;

try
{
    some = Load();
}
catch (Exception)
{
    if (!ignoreFailures) throw;
}

if (some != null)
{
    Use(some);
}

```

Reply:
With slight modification to your pattern, the two possible approaches can be simplified to this:

**1. Use `some` object to track its own loading status:**

```
object some = new object();

try
{
    some = Load();
}
catch (Exception)
{
    some = null;    // this looks sneaky, considering it was never null to begin with
    if (canLoadDifferently())
        some = LoadDifferently();
    if (!ignoreFailures && some == null) throw;
}

if (some != null) // this check unintentionally hides the fact that some can be null only if there is an exception
{
    Use(some);
}
```

**2. Use a bool flag to track `some` object's loading status:**

```
// this could introduce a bug in future if the flag is not in sync with the loading status
bool isLoadComplete = false;  
object some = new object();

try
{
    some = Load();
    isLoadComplete = true;
}
catch (Exception)
{
    if (canLoadDifferently())
    {
         some = LoadDifferently();
         isLoadComplete = true;
    }
    if (!ignoreFailures && !isLoadComplete) throw;
}

if (isLoadComplete)
{
    Use(some);
}
```

I can see the downsides of both approaches, and I am not strongly in favor of one over the other. However, I'd like to go with the second approach because it's slightly more readable.

Reply:
Your first example has a problem. The object `some` should be initialized to null. Yes, it would be odd if we reset it to null

Reply:
But in our LoadAll() method, we never initialize `data` to null.
This is what we have:

`IDictionary<string, ConfigurationSetting> data = new Dictionary<string, ConfigurationSetting>(StringComparer.OrdinalIgnoreCase);`

Reply:
We can just change it to initialize to null.

Reply:
That would trigger changes in both `LoadAll` and `LoadKeyValuesRegisteredForRefresh`, and an additional overhead of null checks on `data` in various places.

Reply:
Not really, just need to have `serverData` inside of the `try` block. Just like you have `cacheData` in the `catch` block.

`var serverData = new Dictionary<string, ConfigurationSetting>(StringComparer.OrdinalIgnoreCase);`
.
.
.
data = serverData;

Reply:
or move everything inside the try to get data into it's own method.
```
data = null;
try
{
    data = LoadFromServer();
}
```

Reply:
In that case, we'll have something like this:

```
object some = null;

try
{
    object serverSome = new object();
    serverSome = Load();
    some = serverSome;
}
catch (Exception)
{
    some = null; 
    if (canLoadDifferently())
        some = LoadDifferently();
    if (!ignoreFailures && some == null) throw;
}

if (some != null)
{
    Use(some);
}

```

Reply:
That example seems to have some unnecessary pieces? The logic can be simplified to

```
object some = null;

try
{
    some = Load();
}
catch (Exception)
{
    if (canLoadDifferently())
        some = LoadDifferently();
    if (!ignoreFailures && some == null) throw;
}

if (some != null)
{
    Use(some);
}
```

Reply:
Lets say we get an exception after loading a few key-values from AppConfig. At this point, data will not be null because it has some key-values that were loaded successfully.

Lets say offline cache is not configured and ignoreFailures is false.

That means, we will end up hitting `Use(some)`, and load partial config for the user without throwing any errors.